src/imageops/sample.rs

```diff
@@ -430,15 +430,26 @@ pub fn interpolate_bilinear<P: Pixel>(
     let mut out = samples[0];
     for (i, c) in out.channels_mut().iter_mut().enumerate() {
         let v = wff * sff[i] + wfc * sfc[i] + wcf * scf[i] + wcc * scc[i];
-        // this rounding may introduce quantization errors,
-        // but cannot do anything about it.
-        *c = <P::Subpixel as NumCast>::from(v.round()).unwrap_or({
-            if v < 0.0 {
-                P::Subpixel::DEFAULT_MIN_VALUE
-            } else {
-                P::Subpixel::DEFAULT_MAX_VALUE
-            }
-        })
+        let convert = |v: f32| {
+            <P::Subpixel as NumCast>::from(v).unwrap_or({
+                if v < 0.0 {
+                    P::Subpixel::DEFAULT_MIN_VALUE
+                } else {
+                    P::Subpixel::DEFAULT_MAX_VALUE
+                }
+            })
+        };
+        let new_val = convert(v);
+        // super hacky check to see if primitive is an integer or a float
+        // note that it should always be convertible to f32, so this unwrap is alright.
+        *c = if new_val.to_f32().unwrap() != v {
```

Comment:
`DEFAULT_MAX_VALUE` is defined to be 1.0 if the type is floating point and all integer types have larger values. Perhaps this check could be:
```rust
if P::Subpixel::DEFAULT_MAX_VALUE.to_f32() == 1.0 { ... }
```
That could be resolved at compile-time and thus avoid any branch overhead

(Equality comparisons are usually dangerous for floats, but all small integers can be exactly represented by an f32 so there's no risk here)